### DISCUSSION TITLE
Compilation Error

### DISCUSSION THREAD

**Discussion by gudeh** [The-OpenROAD-Project/OpenROAD#2330] [Q&A] [answered]
Hey guys! I am getting the following compilation error:

> [100%] Linking CXX executable openroad
> cd /home/gudeh/Documents/OpenROAD-flow-scripts/tools/OpenROAD/build/src && /usr/local/bin/cmake -E cmake_link_script CMakeFiles/openroad.dir/link.txt --verbose=1
> /usr/bin/c++  -O3 -DNDEBUG  -Wl,--export-dynamic -rdynamic CMakeFiles/openroad.dir/Design.cc.o CMakeFiles/openroad.dir/Tech.cc.o CMakeFiles/openroad.dir/OpenRoad.cc.o CMakeFiles/openroad.dir/Main.cc.o  -o openroad ifp/src/ifp.a openroad_swig.a gpl/gpl.a dpl/dpl.a dpo/dpo.a fin/fin.a rsz/src/rsz.a ppl/ppl.a stt/stt.a dbSta/src/dbSta.a ../../src/sta/app/libOpenSTA.a odb/src/db/libodb.a odb/src/swig/tcl/odbtcl.a rcx/src/rcx.a rmp/src/rmp.a cts/src/cts.a grt/grt.a tap/src/tap.a gui/gui.a drt/drt.a dst/dst.a mpl/mpl.a mpl2/mpl2.a psm/src/psm.a ant/src/ant.a par/par.a utl/utl.a pdn/src/pdn.a ../third-party/abc/libabc.a /usr/lib/x86_64-linux-gnu/libtcl.so -lpthread /usr/lib/x86_64-linux-gnu/libtclreadline.so /usr/lib/x86_64-linux-gnu/libz.so _openroad_swig_py.a odb/src/swig/python/_odbpy.a ifp/src/_ifp_py.a /usr/lib/x86_64-linux-gnu/librt.so ppl/src/munkres/libMunkres.a -lm -ldl -lrt rsz/src/rsz.a grt/grt.a ant/src/ant.a grt/grt.a ant/src/ant.a dpl/dpl.a grt/src/fastroute/libFastRoute4.1.a stt/stt.a sta/sta_swig.a /usr/local/lib/libboost_serialization.a /usr/local/lib/libboost_system.a /usr/local/lib/libboost_thread.a /usr/lib/gcc/x86_64-linux-gnu/9/libgomp.so /usr/lib/x86_64-linux-gnu/libpthread.so /usr/local/lib/libboost_serialization.a mpl/src/ParquetFP/libParquetFP.a par/src/MLPart/libMLPart.a dbSta/src/dbSta.a gui/gui.a /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5.12.8 /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5.12.8 /usr/lib/x86_64-linux-gnu/libQt5Core.so.5.12.8 ../../src/sta/app/libOpenSTA.a odb/src/db/libodb.a odb/src/defout/libdefout.a odb/src/lefout/liblefout.a odb/src/tm/libtm.a /usr/lib/x86_64-linux-gnu/libtcl.so odb/src/cdl/libcdl.a odb/src/swig/common/libodb_swig_common.a odb/src/defin/libdefin.a odb/src/def/libdef.a odb/src/def/libdefzlib.a odb/src/lefin/liblefin.a odb/src/zutil/libzutil.a utl/utl.a /usr/local/lib/libspdlog.a -lpthread odb/src/lef/liblef.a odb/src/lef/liblefzlib.a /usr/lib/x86_64-linux-gnu/libz.so /usr/lib/x86_64-linux-gnu/libpython3.8.so 
> /usr/bin/ld: drt/drt.a(TritonRoute.cpp.o): in function \`boost::archive::basic_binary_iarchive<fr::frIArchive>::init()':
> TritonRoute.cpp:(.text._ZN5boost7archive21basic_binary_iarchiveIN2fr10frIArchiveEE4initEv[_ZN5boost7archive21basic_binary_iarchiveIN2fr10frIArchiveEE4initEv]+0xc4): undefined reference to `boost::archive::detail::basic_iarchive::set_library_version(boost::archive::library_version_type)'
> collect2: error: ld returned 1 exit status
> make[2]: *** [src/CMakeFiles/openroad.dir/build.make:210: src/openroad] Error 1
> make[2]: Leaving directory '/home/gudeh/Documents/OpenROAD-flow-scripts/tools/OpenROAD/build'
> make[1]: *** [CMakeFiles/Makefile2:449: src/CMakeFiles/openroad.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> [100%] Linking CXX executable trTest
> cd /home/gudeh/Documents/OpenROAD-flow-scripts/tools/OpenROAD/build/src/drt && /usr/local/bin/cmake -E cmake_link_script CMakeFiles/trTest.dir/link.txt --verbose=1
> /usr/bin/c++  -O3 -DNDEBUG   CMakeFiles/trTest.dir/test/gcTest.cpp.o CMakeFiles/trTest.dir/test/fixture.cpp.o CMakeFiles/trTest.dir/test/stubs.cpp.o CMakeFiles/trTest.dir/__/gui/src/stub.cpp.o  -o trTest drt.a ../odb/src/db/libodb.a ../stt/stt.a ../dst/dst.a ../sta/sta_swig.a /usr/local/lib/libboost_serialization.a /usr/local/lib/libboost_system.a /usr/local/lib/libboost_thread.a ../dbSta/src/dbSta.a ../gui/gui.a ../odb/src/db/libodb.a ../odb/src/cdl/libcdl.a ../odb/src/defin/libdefin.a ../odb/src/def/libdef.a ../odb/src/def/libdefzlib.a ../odb/src/defout/libdefout.a ../odb/src/lefin/liblefin.a ../odb/src/lef/liblef.a ../odb/src/lef/liblefzlib.a ../odb/src/lefout/liblefout.a ../odb/src/tm/libtm.a ../odb/src/zutil/libzutil.a /usr/lib/x86_64-linux-gnu/libtcl.so /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5.12.8 /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5.12.8 /usr/lib/x86_64-linux-gnu/libQt5Core.so.5.12.8 ../../../src/sta/app/libOpenSTA.a ../utl/utl.a /usr/local/lib/libspdlog.a -lpthread /usr/lib/gcc/x86_64-linux-gnu/9/libgomp.so /usr/lib/x86_64-linux-gnu/libpthread.so /usr/local/lib/libboost_serialization.a /usr/lib/x86_64-linux-gnu/libz.so 
> make[2]: Leaving directory '/home/gudeh/Documents/OpenROAD-flow-scripts/tools/OpenROAD/build'
> [100%] Built target trTest
> make[1]: Leaving directory '/home/gudeh/Documents/OpenROAD-flow-scripts/tools/OpenROAD/build'
> make: *** [Makefile:133: all] Error 2

The error I pasted here is from a clean clone from the flow repository. I did the following commands to get to this error:

> git clone --recursive https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts
> cd OpenROAD-flow-scripts
> sudo ./tools/OpenROAD/etc/DependencyInstaller.sh -dev
> ./build_openroad.sh --local

I noticed that if I run DepdencyInstaller multiple times it actually updates and compile Boost everytime. Not sure how to proceed, any help is welcome.

**Comment by maliberty** [comment]
What OS and what compiler?

**Reply by gudeh** [reply]
> gudeh@snowflake:\~/Documents/OpenROAD-flow-scripts$ uname -ra
>Linux snowflake 5.15.0-48-generic #54\~20.04.1-Ubuntu SMP Thu Sep 1 16:17:26 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux


Sorry, I am sending the complete compilation log. 
[log.txt](https://github.com/The-OpenROAD-Project/OpenROAD/files/9716414/log.txt)

**Reply by vvbandeira** [reply]
Make sure your boost install has all the required features. If you installed from source it should look something like [here](https://github.com/The-OpenROAD-Project/OpenROAD/blob/548b7413073245e1480810a8685075600ac1fc6e/etc/DependencyInstaller.sh#L56)

**Comment by maliberty** [comment]
The log unfortunately doesn't have the boost version so I just added print that in the OR master branch.  What version of boost do you have installed?

**Reply by gudeh** [reply]
Reuploading the log from compiling and from calling DependencyInstaller, after pulling your last udpate: 

[log.txt](https://github.com/The-OpenROAD-Project/OpenROAD/files/9717584/log.txt)
[logDependencyInstaller.txt](https://github.com/The-OpenROAD-Project/OpenROAD/files/9717585/logDependencyInstaller.txt)

**Comment by gudeh** [comment, marked as answer]
Hi everyone, I was able to solve the issue by manually updating Boost. I used the links from [this stackoverflow page](https://stackoverflow.com/questions/2829523/upgrading-boost-version) and set the "-prefix=" with the same location my Boost was already installed. Also, I had to clean my previous build from OR.

It seems that for some reason the DependencyInstaller didn't fetch the correct Boost version to upgrade to.

**Comment by maliberty** [comment]
@vvbandeira any idea about " for some reason the DependencyInstaller didn't fetch the correct Boost version to upgrade to."?

**Reply by vvbandeira** [reply]
@gudeh could you clarify the statement? What was the expected version that it should have fetched and which it did fetch? Note that from the log files you provided before, the script is working as intended -- ie., downloading and installing boost 1.76.0; Finally, if you had already boost installed in your system in a place other than `/usr/local/` I would file this as a corner case that our script is not intended to handle.

**Reply by gudeh** [reply]
Hi @vvbandeira. I said that assuming that 1.76 was not the correct version, if this is the correct version you can ignore what I said. I agree with you, this looks like a corner case and for some reason updating Boost fixed the issue.
